backtracking/graph_coloring.cpp

```diff
@@ -1,72 +1,107 @@
+        for (int i = 0; i < V; i++) {
+            if (graph[v][i] && c == color[i]) {
+                return false;
+            }
+        }
+        return true;
     }
 
-    /* Consider this vertex v and try different colors */
-    for (int c = 1; c <= m; c++) {
-        /* Check if assignment of color c to v is fine*/
-        if (isSafe(v, graph, color, c)) {
-            color[v] = c;
+    /** A recursive utility function to solve m coloring problem
+     * @tparam V number of vertices in the graph
+     * @param graph description
+     * @param m description
+     * @param [in,out] color description // used in,out to notify in documentation
+     * that this parameter gets modified by the function
+     * @param v description
+     */
+    template <size_t V>
+    void graphColoring(const std::array<std::array <int, V>, V>& graph, int m, std::array <int, 4> color, int v) {
+        // base case:
+        // If all vertices are assigned a color then return true
+        if (v == V) {
+            backtracking::printSolution(color, V);
+            return;
+        }
 
-            /* recur to assign colors to rest of the vertices */
-            graphColoring(graph, m, color, v + 1);
+        // Consider this vertex v and try different colors
+        for (int c = 1; c <= m; c++) {
+            // Check if assignment of color c to v is fine
+            if (backtracking::isSafe<V>(v, graph, color, c)) {
+                color[v] = c;
 
-            /* If assigning color c doesn't lead to a solution
-               then remove it */
-            color[v] = 0;
+                // recur to assign colors to rest of the vertices
+                backtracking::graphColoring<V>(graph, m, color, v + 1);
+
+                // If assigning color c doesn't lead to a solution then remove it
+                color[v] = 0;
+            }
         }
     }
-}
-
-/* A utility function to print solution */
-void printSolution(int color[]) {
-    printf(" Following are the assigned colors \n");
-    for (int i = 0; i < V; i++) printf(" %d ", color[i]);
-    printf("\n");
-}
+}  // namespace backtracking
 
-// driver program to test above function
+/**
+ * Main function
+ */
 int main() {
-    /* Create following graph and test whether it is 3 colorable
-      (3)---(2)
-       |   / |
-       |  /  |
-       | /   |
-      (0)---(1)
-    */
-    bool graph[V][V] = {
-        {0, 1, 1, 1},
-        {1, 0, 1, 0},
-        {1, 1, 0, 1},
-        {1, 0, 1, 0},
+    // Create following graph and test whether it is 3 colorable
+    // (3)---(2)
+    // |   / |
+    // |  /  |
+    // | /   |
+    // (0)---(1)
+
+    const int V = 4;  // number of vertices in the graph
+    std::array <std::array <int, V>, V> graph = {
+        {0, 1, 1, 1,
+        1, 0, 1, 0,
+        1, 1, 0, 1,
+        1, 0, 1, 0}
     };
     int m = 3;  // Number of colors
 
-    int color[V];
+    std::array <int, 4> color;
```

Comment:
```suggestion
    std::array <int, V> color{};
```
This initialization would set all elements to `0` by default without the need for the loop on line 101